HHH-13410 : "order_inserts = true" causes FK Violation when inserting…

nucleussoftwareopen opened the pull request:
https://hibernate.atlassian.net/browse/HHH-13410

This issue is due to below code in ActionQueue as nextBatchIdentifier.parent is assigned a parent. There can be multiple parents but with this code only one parent reference is being hold.  
```java 
for ( int j = i + 1; j < latestBatches.size(); j++ ) {
                    BatchIdentifier nextBatchIdentifier = latestBatches.get( j );

                    if ( nextBatchIdentifier.hasAnyParentEntityNames( batchIdentifier ) ) {
                        nextBatchIdentifier.parent = batchIdentifier;
                                           }
                    if ( batchIdentifier.hasAnyChildEntityNames( nextBatchIdentifier ) ) {
                        nextBatchIdentifier.parent = batchIdentifier;
                     }
                }
```
Fix:  parent should be added to nextBatchIdentifier.parentEntityNames

```java
                for ( int j = i + 1; j < latestBatches.size(); j++ ) {
                    BatchIdentifier nextBatchIdentifier = latestBatches.get( j );

                    if ( nextBatchIdentifier.hasAnyParentEntityNames( batchIdentifier ) ) {
                        nextBatchIdentifier.parent = batchIdentifier;
                        nextBatchIdentifier.getParentEntityNames().add(batchIdentifier.getEntityName());
                    }
                    if ( batchIdentifier.hasAnyChildEntityNames( nextBatchIdentifier ) ) {
                        nextBatchIdentifier.parent = batchIdentifier;
                        nextBatchIdentifier.getParentEntityNames().add(batchIdentifier.getEntityName());
                    }
                }
```

NathanQingyangXu commented:
Seems a good bug reporting. However, you need to add some testing case to showcase:

1. without your code changes, the testing case would fail as reported in the ticket;
2. without your code changes, the testing case passes.

Please refer to existing hibernate testing case for how to create the testing case.

NathanQingyangXu commented on hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java:
```diff
@@ -1189,9 +1189,11 @@ public void sort(List<AbstractEntityInsertAction> insertions) {
 					}
 					if ( batchIdentifier.hasAnyChildEntityNames( nextBatchIdentifier ) ) {
 						nextBatchIdentifier.parent = batchIdentifier;
+						nextBatchIdentifier.getParentEntityNames().add(batchIdentifier.getEntityName());
```
Please refer to https://github.com/hibernate/hibernate-ide-codestyles on how to make your code style consistent with existing one. Especially, empty chars are added aggressively to help code reading. For instance
```
nextBatchIdentifier.getParentEntityNames().add(batchIdentifier.getEntityName());
``` 
should be re-formatted as
```
nextBatchIdentifier.getParentEntityNames().add( batchIdentifier.getEntityName() );
```

nucleussoftwareopen replied:
updated

nucleussoftwareopen commented:
@NathanQingyangXu I've updated code as per your review, also added test case which is passing with this fix.
Without fix ConstraintViolationException is being thrown.

Sanne commented:
thanks @nucleussoftwareopen and @NathanQingyangXu ! Merging..

gavinking commented:
Replicated to HR in https://github.com/hibernate/hibernate-reactive/commit/477e421476403d179d37615e7622040f682c75c3.
